File: infrastructures/ddd/python/src/application/errors/infra/service_exception.py
"""
Base class for all service-level exceptions.
All service exceptions MUST inherit from this class.
Location: MUST be in /application/errors/infra/ directory
"""

from abc import ABC, abstractmethod
from enum import Enum
from typing import Dict, Generic, Optional, Type, TypeVar


TErrorCode = TypeVar('TErrorCode', bound=Enum)


class ServiceException(Exception, ABC):
    """
    Base class for all service-level exceptions.
    
    All service exceptions MUST:
    - Inherit from this class
    - Define an error code enum
    - Use typed ServiceException[ErrorCodeEnum]
    - Live in /application/errors/
    - Follow pattern: {Object}ServiceException
    """
    
    @property
    @abstractmethod
    def generic_error_code(self) -> Enum:
        """Gets the generic error code as an Enum."""
        pass
    
    @property
    @abstractmethod
    def error_code_type(self) -> Type[Enum]:
        """Gets the type of the error code enum."""
        pass


class TypedServiceException(ServiceException, Generic[TErrorCode]):
    """
    Generic base class for typed service exceptions.
    
    Pattern: Create {Object}ServiceException(TypedServiceException[{Object}ErrorCode])
    Location: MUST be in /application/errors/ directory
    
    Usage:
        class ExampleErrorCode(Enum):
            NOT_FOUND = "not_found"
            VALIDATION_FAILED = "validation_failed"
        
        class ExampleServiceException(TypedServiceException[ExampleErrorCode]):
            _MESSAGE_TEMPLATES = {
                ExampleErrorCode.NOT_FOUND: "Example '{id}' not found",
                ExampleErrorCode.VALIDATION_FAILED: "Validation failed: {reason}"
            }
            
            def __init__(
                self,
                p_code: ExampleErrorCode,
                p_details: Optional[Dict[str, any]] = None
            ):
                super().__init__(p_code, self._MESSAGE_TEMPLATES, p_details)
    """
    
    def __init__(
        self,
        p_code: TErrorCode,
        p_message_templates: Dict[Enum, str],
        p_details: Optional[Dict[str, any]] = None
    ):
        """
        Initialize a typed service exception.
        
        Args:
            p_code: The specific error code
            p_message_templates: Dictionary mapping error codes to message templates
            p_details: Optional structured details about the error
        """
        self._error_code = p_code
        self._details = p_details or {}
        
        message = self._format_message(p_code, p_message_templates, self._details)
        super().__init__(message)
    
    @property
    def error_code(self) -> TErrorCode:
        """The specific error code for this exception."""
        return self._error_code
    
    @property
    def details(self) -> Dict[str, any]:
        """Optional structured details about the error."""
        return self._details
    
    @property
    def generic_error_code(self) -> Enum:
        """Gets the generic error code as an Enum."""
        return self._error_code
    
    @property
    def error_code_type(self) -> Type[Enum]:
        """Gets the type of the error code enum."""
        return type(self._error_code)
    
    @staticmethod
    def _format_message(
        p_code: Enum,
        p_message_templates: Dict[Enum, str],
        p_details: Dict[str, any]
    ) -> str:
        """
        Formats the exception message using the template and details.
        
        Args:
            p_code: The error code
            p_message_templates: Dictionary of message templates
            p_details: Dictionary of detail values
            
        Returns:
            The formatted message string
        """
        template = p_message_templates.get(p_code)
        
        if template is None:
            return f"Error occurred: {p_code.name}"
        
        if not p_details:
            return template
        
        message = template
        for key, value in p_details.items():
            message = message.replace(f"{{{key}}}", str(value))
        
        return message

File: infrastructures/ddd/typescript/src/application/errors/infra/ServiceException.ts
/**
 * Base class for all service exceptions.
 * Service exceptions:
 * - Live in /application/errors/ (NOT /services/)
 * - Suffix with "ServiceException"
 * - Pattern: {Object}ServiceException extends ServiceException<{Object}ErrorCode>
 * - Define error codes as enums (NOT strings)
 * - Include message templates for formatting
 * - Include optional details object
 * 
 * Error codes map to HTTP status codes:
 * - NotFound → 404
 * - Conflict → 409
 * - ValidationFailed → 400
 * - Unauthorized → 401
 * - Forbidden → 403
 * - Unknown → 500
 * 
 * Example usage:
 * throw new ExampleServiceException(
 *   ExampleErrorCode.NotFound,
 *   { id: exampleId }
 * );
 */
export abstract class ServiceException<TErrorCode> extends Error {
  public readonly ErrorCode: TErrorCode;
  public readonly Details?: Record<string, unknown>;

  protected constructor(
    p_errorCode: TErrorCode,
    p_messageTemplates: Record<string, string>,
    p_details?: Record<string, unknown>
  ) {
    const message = ServiceException.FormatMessage(
      p_errorCode,
      p_messageTemplates,
      p_details
    );
    super(message);
    
    this.ErrorCode = p_errorCode;
    this.Details = p_details;
    this.name = this.constructor.name;

    // Maintains proper stack trace for where our error was thrown (only available on V8)
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, this.constructor);
    }
  }

  /**
   * Formats the error message using the template and details.
   */
  private static FormatMessage<TErrorCode>(
    p_errorCode: TErrorCode,
    p_messageTemplates: Record<string, string>,
    p_details?: Record<string, unknown>
  ): string {
    const codeKey = String(p_errorCode);
    let template = p_messageTemplates[codeKey] || 'An error occurred';

    if (p_details) {
      // Replace placeholders like {id} with actual values from details
      for (const [key, value] of Object.entries(p_details)) {
        template = template.replace(`{${key}}`, String(value));
      }
    }

    return template;
  }

  /**
   * Gets the generic error code as a string.
   * Used by error handling middleware.
   */
  public get GenericErrorCode(): string {
    return String(this.ErrorCode);
  }
}
